CHECK_CXX_SOURCE_COMPILES does not use the correct compile flags.

@tamiko This is an issue I need your help with.

The underlying cause I'm trying to resolve is that libc++ (clang's C++ library) does not implement `std::cyl_bessel_j`. We test this with the following piece of code in `cmake/checks/check_01_cxx_features`:
```
#
# Check for c++17 Bessel function support. Unfortunately libc++ version 10
# does not have those.
#

CHECK_CXX_SOURCE_COMPILES(
  "
  #include <cmath>
  using std::cyl_bessel_j;
  using std::cyl_bessel_jf;
  int main()
  {
  }
  "
  DEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS
  )
```

The problem is the following. Let's say I do this:
```
export CXXFLAGS="-stdlib=libc++"
export CXX=clang++
cmake ../dealii
```
Then if you go to `CMakeFiles/CMakeConfigureLog.yaml` and search for BESSEL, you will find an entry like this:
```
    kind: "try_compile-v1"
    backtrace:
      - "/usr/share/cmake-3.28/Modules/Internal/CheckSourceCompiles.cmake:101 (try_compile)"
      - "/usr/share/cmake-3.28/Modules/CheckCXXSourceCompiles.cmake:52 (cmake_check_source_compiles)"
      - "cmake/checks/check_01_cxx_features.cmake:615 (CHECK_CXX_SOURCE_COMPILES)"
      - "cmake/macros/macro_verbose_include.cmake:18 (include)"
      - "CMakeLists.txt:123 (verbose_include)"
    checks:
      - "Performing Test DEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS"
    directories:
      source: "/home/bangerth/p/deal.II/2/build-test/CMakeFiles/CMakeScratch/TryCompile-OTSmON"
      binary: "/home/bangerth/p/deal.II/2/build-test/CMakeFiles/CMakeScratch/TryCompile-OTSmON"
    cmakeVariables:
      CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS: "/usr/bin/clang-scan-deps-18"
      CMAKE_CXX_FLAGS: ""
      CMAKE_CXX_FLAGS_DEBUG: ""
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "/home/bangerth/p/deal.II/2/cxx-modules/cmake/modules/"
    buildResult:
      variable: "DEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS"
      cached: true
      stdout: |
        Change Dir: '/home/bangerth/p/deal.II/2/build-test/CMakeFiles/CMakeScratch/TryCompile-OTSmON'
        
        Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_929cc/fast
        /usr/bin/gmake  -f CMakeFiles/cmTC_929cc.dir/build.make CMakeFiles/cmTC_929cc.dir/build
        gmake[1]: Entering directory '/home/bangerth/p/deal.II/2/build-test/CMakeFiles/CMakeScratch/TryCompile-OTSmON'
        Building CXX object CMakeFiles/cmTC_929cc.dir/src.cxx.o
        /usr/bin/clang++ -DDEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS  -pedantic -Werror -Wno-unused-command-line-argument -std=c++17 -MD -MT CMakeFiles/cmTC_929cc.dir/src.cxx.o -MF CMakeFiles/cmTC_929cc.dir/src.cxx.o.d -o CMakeFiles/cmTC_929cc.dir/src.cxx.o -c /home/bangerth/p/deal.II/2/build-test/CMakeFiles/CMakeScratch/TryCompile-OTSmON/src.cxx
        Linking CXX executable cmTC_929cc
        /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_929cc.dir/link.txt --verbose=1
        /usr/bin/clang++  -pedantic -Werror -Wno-unused-command-line-argument CMakeFiles/cmTC_929cc.dir/src.cxx.o -o cmTC_929cc 
        gmake[1]: Leaving directory '/home/bangerth/p/deal.II/2/build-test/CMakeFiles/CMakeScratch/TryCompile-OTSmON'
        
      exitCode: 0
```
So it detects the presence of `std::cyl_bessel_j`, *but that is because it calls Clang without the `-stdlib=libc++` flag*. So it looks for the presence of that function in libstdc++ (the GCC library) where it is correctly implemented, even though we want to know whether it's in libc++.

I have no idea where this needs to be addressed, but it's clearly a bug in our detection machinery.

@bangerth With #18250 
```
  -
    kind: "try_compile-v1"
    backtrace:
      - "/usr/share/cmake/Modules/Internal/CheckSourceCompiles.cmake:108 (try_compile)"
      - "/usr/share/cmake/Modules/CheckCXXSourceCompiles.cmake:58 (cmake_check_source_compiles)"
      - "cmake/checks/check_01_cxx_features.cmake:572 (CHECK_CXX_SOURCE_COMPILES)"
      - "cmake/macros/macro_verbose_include.cmake:18 (include)"
      - "CMakeLists.txt:123 (verbose_include)"
    checks:
      - "Performing Test DEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS"
    directories:
      source: "/home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg"
      binary: "/home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg"
    cmakeVariables:
      CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS: "/usr/lib/llvm/20/bin/clang-scan-deps"
      CMAKE_CXX_FLAGS: ""
      CMAKE_CXX_FLAGS_DEBUG: ""
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "/home/tamiko/workspace/dealii/cmake/modules/"
    buildResult:
      variable: "DEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS"
      cached: true
      stdout: |
        Change Dir: '/home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg'
        
        Run Build Command(s): /usr/bin/ninja -v cmTC_56f27
        [1/2 | 00:00 (ETA 00:00)] /usr/lib/llvm/20/bin/clang++ -DDEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS  -openmp-simd -stdlib=libc++ -pedantic -Werror -Wno-unused-command-line-argument -std=c++17 -MD -MT CMakeFiles/cmTC_56f27.dir/src.cxx.o -MF CMakeFiles/cmTC_56f27.dir/src.cxx.o.d -o CMakeFiles/cmTC_56f27.dir/src.cxx.o -c /home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg/src.cxx
        FAILED: CMakeFiles/cmTC_56f27.dir/src.cxx.o 
        /usr/lib/llvm/20/bin/clang++ -DDEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS  -openmp-simd -stdlib=libc++ -pedantic -Werror -Wno-unused-command-line-argument -std=c++17 -MD -MT CMakeFiles/cmTC_56f27.dir/src.cxx.o -MF CMakeFiles/cmTC_56f27.dir/src.cxx.o.d -o CMakeFiles/cmTC_56f27.dir/src.cxx.o -c /home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg/src.cxx
        /home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg/src.cxx:3:14: error: no member named 'cyl_bessel_j' in namespace 'std'
            3 |   using std::cyl_bessel_j;
              |         ~~~~~^
        /home/tamiko/workspace/dealii/build/CMakeFiles/CMakeScratch/TryCompile-bzXErg/src.cxx:4:14: error: no member named 'cyl_bessel_jf' in namespace 'std'
            4 |   using std::cyl_bessel_jf;
              |         ~~~~~^
        2 errors generated.
        ninja: build stopped: subcommand failed.
        
      exitCode: 1
```

--- a/cmake/checks/check_01_cxx_features.cmake
+++ b/cmake/checks/check_01_cxx_features.cmake
@@ -36,7 +36,6 @@
 #                                                                      #
 ########################################################################
 
-
 #
 # We need compiler flags specified in ${DEAL_II_CXX_FLAGS} for all the
 # tests. Create a small macro to easily set CMAKE_REQUIRED_FLAGS
@@ -413,6 +412,12 @@ else()
   enable_if_supported(_werror_flag "-Werror")
   enable_if_supported(_werror_flag "-Wno-unused-command-line-argument")
 endif()
+
+#
+# We need to reset CMAKE_REQUIRED_* variables again after the above call to
+# enable_if_supported()
+#
+_set_up_cmake_required()
 add_flags(CMAKE_REQUIRED_FLAGS "${_werror_flag}")
 
 unset_if_changed(CHECK_CXX_FEATURES_FLAGS_SAVED

--- a/cmake/macros/macro_enable_if_links.cmake
+++ b/cmake/macros/macro_enable_if_links.cmake
@@ -14,15 +14,16 @@
 
 #
 # Tests whether it is possible to compile and link a dummy program with a
-# given flag.
-# If so, add it to variable.
+# given flag. If so, add it to variable.
+#
+# Note: This macro will reset the CMAKE_REQUIRED_* variables.
 #
 # Usage:
 #     enable_if_links(variable flag)
 #
 
 macro(enable_if_links _variable _flag)
-  # keep on top to avoid cluttering the _flag and _flag_stripped variables
+  reset_cmake_required()
   enable_if_supported(CMAKE_REQUIRED_FLAGS "-Werror")
 
   string(STRIP "${_flag}" _flag_stripped)
@@ -32,11 +33,12 @@ macro(enable_if_links _variable _flag)
 
     list(APPEND CMAKE_REQUIRED_LIBRARIES "${_flag_stripped}")
     CHECK_CXX_SOURCE_COMPILES("int main(){}" DEAL_II_HAVE_LINKER_FLAG_${_flag_name})
-    reset_cmake_required()
 
     if(DEAL_II_HAVE_LINKER_FLAG_${_flag_name})
       set(${_variable} "${${_variable}} ${_flag_stripped}")
       string(STRIP "${${_variable}}" ${_variable})
     endif()
   endif()
+
+  reset_cmake_required()
 endmacro()

--- a/cmake/macros/macro_enable_if_supported.cmake
+++ b/cmake/macros/macro_enable_if_supported.cmake
@@ -13,21 +13,24 @@
 ## ------------------------------------------------------------------------
 
 #
-# Tests whether the cxx compiler understands a flag.
-# If so, add it to 'variable'.
+# Tests whether the cxx compiler understands a flag. If so, add it to
+# 'variable'.
+#
+# Note: This macro will reset the CMAKE_REQUIRED_* variables.
 #
 # Usage:
 #     enable_if_supported(variable flag)
 #
 
 macro(enable_if_supported _variable _flag)
-  # First check if we can use -Werror
+  reset_cmake_required()
+
+  #
+  # Add -Werror if available:
+  #
   CHECK_CXX_COMPILER_FLAG("-Werror" DEAL_II_HAVE_FLAG_Werror)
-
-  string(STRIP "${_flag}" _flag_stripped)
-
   if(DEAL_II_HAVE_FLAG_Werror)
-    set(CMAKE_REQUIRED_FLAGS "-Werror")
+    string(STRIP "${CMAKE_REQUIRED_FLAGS} -Werror" CMAKE_REQUIRED_FLAGS)
   endif()
 
   #
@@ -37,6 +40,7 @@ macro(enable_if_supported _variable _flag)
   # compilation unit.
   # Therefore we invert the test for -Wno-... flags:
   #
+  string(STRIP "${_flag}" _flag_stripped)
   set(_flag_sanitized "${_flag_stripped}")
   if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
     string(REPLACE "-Wno-" "-W" _flag_sanitized "${_flag_stripped}")
@@ -54,5 +58,5 @@ macro(enable_if_supported _variable _flag)
     endif()
   endif()
 
-  unset(CMAKE_REQUIRED_FLAGS)
+  reset_cmake_required()
 endmacro()

